metersink/output_odoo.py

```diff
@@ -129,157 +236,323 @@ def get_odoo_user_id(odoo):
+        odoo,
+        "sale.order",
+        mode="records",
+        filter_list=filter_list,
+        projection_dict=projection_dict,
+    )
+    if not sale_orders:
+        line_dict = {
+            "partner_id": 75,
+            "client_order_ref": project_id,
+            "require_payment": False,
+            "require_signature": False,
+            "tag_ids": ["cloud", project_id],
+        }
+
+        sale_order_id = odoo_create(odoo, "sale.order", [line_dict])
+
+        end_date = datetime.now()
+        time_calc = calculate_cloud_time(
+            data_dict["traits"]["created_at"], end_date
+        )
+        value_list = [str(data_dict["traits"]["size"])]
+
+        info_dict = {
+            "uuid": data_dict["traits"]["resource_id"],
+            "name": data_dict["traits"]["display_name"],
+            "values": value_list,
+            "start": data_dict["traits"]["created_at"],
+            "end": end_date,
+        }
+        display_name = get_name_from_info(info_dict)
+
+        line_dict = {
+            "product_id": 36,
+            "display_name": display_name,
+            "product_uom_qty": time_calc,
+            "order_id": sale_order_id,
+        }
+        line_id = odoo_create(odoo, "sale.order.line", [line_dict])
 
-def is_in_odoo(odoo, model="account.analytic.line", line=None, filter_dict=None):
+    else:
+        if LOG.isEnabledFor(logging.DEBUG):
+            LOG.debug("%s", pformat(sale_orders))
+
+        display_name = f"{data_dict['traits']['resource_id']} ({data_dict['traits']['size']}GB)"
+
+        # line_ids = sale_orders['order_line']
+        # projection_dict = get_projection_dict(model="sale.order.line")
+
+        # line_records = odoo_get(odoo, 'sale.order.line',
+        #                         mode='read',
+        #                         filter_list=[line_ids],
+        #                         # projection_dict=projection_dict
+        #                         )
+
+        line_record = odoo_get(
+            odoo,
+            "sale.order.line",
+            mode="search_read",
+            filter_list=[
+                [
+                    ["order_id", "=", sale_orders["id"]],
+                    ["product_id", "=", 36],
+                ]
+            ],
+            # projection_dict=projection_dict
+        )
+        if line_record:
+            # todo update
+            # we are reading the last state of the line and calculate the time between this line and the update message.
+
+            if LOG.isEnabledFor(logging.DEBUG):
+                LOG.debug("%s", pformat(line_record))
+
+            time_calc = calculate_cloud_time(line_record["name"])
+
+            quantity = 3000
+            odoo_update(
+                odoo,
+                "sale.order.line",
+                line_record["id"],
+                {"product_uom_qty": quantity},
+            )
+        else:
+            # If there is no line already for the ressource, create it.
```

Comment:
The update procedure just uses the id of the line record. not the record itself. the update is done on the current record.

Reply:
I don't follow. My point is: it could be racy -- if you check whether something exists, then you create the thing, it could happen that in the meantime, the thing actually got created by some other party, and then you end up with two of these things (if that is possible), or an error, or maybe lost data.